lib/usagereporter/usagereporter.go

```diff
@@ -230,6 +240,11 @@ func (r *UsageReporter[T]) enqueueBatch() {
 	}
 }
 
+// GracefulStop stops receiving new events and schedules a batch for submission.
+func (r *UsageReporter[T]) GracefulStop() {
```

Comment:
I'm not an expert on this but something about adding those two extra channels to coordinate graceful stop doesn't sit right with me.

In general, I'm not sure how to implement graceful shutdown for a struct which has a Run method that accepts a context. I remembered the Slack discussion we had where [Alan said that contexts are request-based](https://gravitational.slack.com/archives/C0DF0TPMY/p1664286378152379?thread_ts=1664211584.686659&cid=C0DF0TPMY) and are best suited for that use case rather than controlling the lifecycle of something bigger than the request. I found another discussion in Go's issue tracker where some people express a similar opinion (https://github.com/golang/go/issues/54775#issuecomment-1233604743).

If GracefulStop is graceful then should closing the context passed to Run stop UsageReporter in a non-graceful manner? I don't know so I'm curious what others think about it.

Reply:
Yeah, I think the best API here is something like this:

```
// GracefulStop stops receiving new events and schedules the
// final batch for submission. It blocks until the final batch
// has been sent, or until the provided context is canceled.
func (f *UsageReporter[T] ) GracefulStop(ctx context.Context) error {}
```

And then if the caller wants to enforce a timeout, they can use `context.WithTimeout` to pass in a context that times out. The returned error would come from `ctx.Err()` and can be used to determine if the context timed out or the stop completed successfully.

Reply:
Alternatively, you could do without all of this, and just add one final enqueue at the end:

```
// Run begins processing incoming usage events. It should be run in a goroutine.
func (r *UsageReporter[T]) Run(ctx context.Context) {
	timer := r.clock.NewTimer(r.maxBatchAge)

	// Also start the submission goroutine.
	go r.runSubmit(ctx)

	r.Debug("usage reporter is ready")

	for {
		select {
		case <-ctx.Done():
			// before returning, submit one final batch
			submitLastBatch()
			return
```

Reply:
I also wonder what happens when we enter cancellation (so the for loop in Run no longer reads from `r.events` which is `make(chan []*SubmittedEvent[T], 1)`) and there are two goroutines calling `AddEventsToQueue` (which ends with `r.events <- submitted`). Wouldn't at least one of them hang indefinitely?

Reply:
If so, maybe we should have a variable `isShutdown`, similarly to [http server](https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/net/http/server.go;l=2773;drc=7026a8a1a9aba58285b3cae215df04140245b6db)? We would drop new events in the shutdown mode.

Reply:
I think adding a variable would mean adding a mutex all the sync issues that entails. Maybe instead we can just discard writes when the channel gets full to avoid hanging?

Reply:
I think it could potentially cause the events to be dropped during normal operation (?)
It seems that mutex is a safer way here.